lib/Dialect/QUIR/Transforms/MergeCircuitMeasures.cpp

```diff
@@ -0,0 +1,434 @@
+                llvm::SmallVector<Value> &outputValues, quir::ReturnOp returnOp,
+                CallCircuitOp callCircuitOp, CircuitOp circuitOp,
+                llvm::StringRef newName1) {
+  rewriter.setInsertionPointAfter(returnOp);
+  auto newReturnOp =
+      rewriter.create<quir::ReturnOp>(returnOp->getLoc(), outputValues);
+  // newReturnOp.dump();
+  rewriter.replaceOp(returnOp, newReturnOp->getResults());
+
+  // change the input / output types for the quir.circuit
+  auto opType = circuitOp.getFunctionType();
+  circuitOp.setType(rewriter.getFunctionType(
+      /*inputs=*/opType.getInputs(),
+      /*results=*/ArrayRef<Type>(outputTypes)));
+
+  // new call circuit ops
+  rewriter.setInsertionPointAfter(callCircuitOp);
+  auto newCallOp = rewriter.create<mlir::quir::CallCircuitOp>(
+      callCircuitOp->getLoc(), newName1, TypeRange(outputTypes),
+      ValueRange(callCircuitOp->getOperands()));
+  return newCallOp;
+}
+
+static void dropNextMeasure(PatternRewriter &rewriter,
+                            llvm::SmallVector<Type> &outputTypes,
+                            llvm::SmallVector<Value> &outputValues,
+                            CircuitOp nextCircuitOp, MeasureOp nextMeasureOp) {
+  auto nextReturnOp = dyn_cast<quir::ReturnOp>(&nextCircuitOp.back().back());
+  assert(nextReturnOp && "quir.circuit must end end a quir.return");
+  outputTypes.clear();
+  outputValues.clear();
+  std::vector<int> eraseList;
+  for (uint idx = 0; idx < nextReturnOp.getNumOperands(); idx++)
+    if (nextReturnOp->getOperand(idx).getDefiningOp() ==
+        nextMeasureOp.getOperation())
+      eraseList.push_back(idx);
+
+  while (!eraseList.empty()) {
+    nextReturnOp->eraseOperand(eraseList.back());
+    eraseList.pop_back();
+  }
+
+  rewriter.eraseOp(nextMeasureOp);
+
+  outputTypes.append(nextReturnOp->getOperandTypes().begin(),
+                     nextReturnOp->getOperandTypes().end());
+
+  auto opType = nextCircuitOp.getFunctionType();
+  nextCircuitOp.setType(rewriter.getFunctionType(
+      /*inputs=*/opType.getInputs(),
+      /*results=*/ArrayRef<Type>(outputTypes)));
+}
+
+static void mergeMeasurements(PatternRewriter &rewriter,
+                              CallCircuitOp callCircuitOp,
+                              CallCircuitOp nextCallCircuitOp,
+                              CircuitOp circuitOp, CircuitOp nextCircuitOp,
+                              MeasureOp measureOp, MeasureOp nextMeasureOp,
+                              llvm::StringMap<Operation *> &symbolMap) {
+
+  // copy circuitOp in case there are multiple calls
+  std::string const newNameTemplate =
+      (circuitOp.getSymName() + "_" + nextCircuitOp.getSymName()).str();
+  auto newName1 =
+      duplicateCircuit(rewriter, circuitOp, symbolMap, newNameTemplate, "+m");
+
+  // copy nextCircuitOp in case there are multiple calls
+  auto newName2 = duplicateCircuit(rewriter, nextCircuitOp, symbolMap,
+                                   newNameTemplate, "-m");
```

Comment:
In general, I'm not sure I follow why we have two circuits in the first place? Shouldn't the circuits have been merged at this point into a single circuit (upon which we could apply the existing measurement merging pass within the circuit)?

Reply:
We have cases where the circuits are not being merged due to `qcs.parallel_control_flow` or switch operations in between two circuits. Without circuits the existing merge-measures is able to merge across the qcs.paralle_control_flow or switch operations for the individual measures. With circuits while the measures could be merged there could be additional operations in the second circuit which would prevent the two circuits from being fully merged due to the control flow. What the new pass is doing is looking for measures which may be moved from one circuit to the second to enable merging. 

I am duplicating circuits because I do not want to assume that the circuits are only called once. By duplicating the circuits any additional calls to the circuit will not be effected by the changes being made to support the measurement merge.

I duplicate both circuits because, while I am moving the measure from one circuit to the other, there may be additional operations in the "source" circuit which should not be moved and should be executed where the original call circuit was in the program flow.

Reply:
Thank you, I was wondering if we should check to make sure the leftover circuits are removed if there are no further callers? This would be to avoid blow-up of circuits (which I would imagine in the worst case would happen with unique circuits on containing a single measure on independent sets of qubits)
Ie., 
```
quir.circuit @qc0 ()... {
   quir.measure(%0)
}
...
quir.circuit @qcN ()... {
   quir.measure(%N)
}


quir.call_circuit @qc0 ()
...
quir.call_circuit @qcN ()
```

Reply:
Right now there is nothing that I'm aware of in the code base to remove unused circuits. I can add a pass to do that.

Reply:
Added `RemoveUnusedCircuitsPass` with 5c93749